python/cudf_polars/cudf_polars/experimental/parallel.py

```diff
@@ -284,12 +284,16 @@ def _(
     ir: Union, partition_info: MutableMapping[IR, PartitionInfo]
 ) -> MutableMapping[Any, Any]:
     key_name = get_key_name(ir)
-    partition = itertools.count()
-    return {
-        (key_name, next(partition)): child_key
-        for child in ir.children
-        for child_key in partition_info[child].keys(child)
-    }
+    if partition_info[ir].count == 1:
+        child_keys = [k for c in ir.children for k in partition_info[c].keys(c)]
+        return {(key_name, 0): (Union.do_evaluate, ir.zlice, *child_keys)}
+    else:
+        partition = itertools.count()
+        return {
+            (key_name, next(partition)): child_key
+            for child in ir.children
+            for child_key in partition_info[child].keys(child)
+        }
```

Comment:
Is the iteration logic the same in both branches? If so, could we consolidate it? (roughly like this):

```suggestion
    child_keys = [
        child_key
        for child in ir.children
        for child_key in partition_info[child].keys(child)
    ]
    if partition_info[ir].count == 1:
        return {(key_name, 0): (Union.do_evaluate, ir.zlice, *child_keys)}
    else:
        return {
            (key_name, i): child_key
            for i, child_key in enumerate(child_keys)
        }
```

I think you'd need to remove the `itertools` import if we do that. We could also keep the `next(partition)` thing if you want.

Reply:
Thanks done